hetmech/matrix.py

```diff
@@ -69,5 +75,43 @@ def normalize(matrix, vector, axis, damping_exponent):
         vector **= -damping_exponent
     vector[numpy.isinf(vector)] = 0
     shape = (len(vector), 1) if axis == 'rows' else (1, len(vector))
-    matrix *= vector.reshape(shape)
+    vector = vector.reshape(shape)
+    if sparse.issparse(matrix):
+        matrix = matrix.multiply(vector)
+    else:
+        matrix *= vector
+    return matrix
+
+
+def auto_convert(matrix, threshold):
+    """
+    Automatically convert a scipy.sparse to a numpy.ndarray if the percent
+    nonzero is above a given threshold. Automatically convert a numpy.ndarray
+    to scipy.sparse if the percent nonzero is below a given threshold.
+
+    Parameters
+    ==========
+    matrix : numpy.ndarray or scipy.sparse
+    threshold : float (0 < threshold < 1)
+        percent nonzero above which the matrix is converted to dense
+
+    Returns
+    =======
+    matrix : numpy.ndarray or scipy.sparse
+    """
+    above_thresh = (matrix != 0).sum() / numpy.prod(matrix.shape) >= threshold
+    if sparse.issparse(matrix) and above_thresh:
+        matrix = matrix.toarray()
```

Comment:
You can go if, if, and then a final return matrix in case neither was triggered. The point is elif is equivalent to if when the previous if clause returns.